src/main/java/dog/pawbook/model/dog/Dog.java

```diff
@@ -0,0 +1,156 @@
+package dog.pawbook.model.dog;
+
+import static dog.pawbook.commons.util.CollectionUtil.requireAllNonNull;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+
+import dog.pawbook.model.Identifiable;
+import dog.pawbook.model.owner.Name;
+import dog.pawbook.model.owner.Owner;
+import dog.pawbook.model.tag.Tag;
+
+
+/**
+ * Represents a dog in the address book.
+ * Guarantees: details are present and not null, field values are validated, immutable.
+ */
+public class Dog implements Identifiable<Dog> {
```

Comment:
I'm not so sure if we want interface templates instead of abstract base class, since interfaces only define methods, no common data like name or tag can be made common, the other thing is when it comes to writing functions like find it might be difficult to avoid boilerplate code.